src/modules/form/components/viewable/DynamicViewFields.tsx

```diff
@@ -0,0 +1,150 @@
+  disabledLinkIds: string[];
+  setDisabledLinkIds: React.Dispatch<React.SetStateAction<string[]>>;
+}
+
+export const isEnabled = (
+  item: FormItem,
+  disabledLinkIds: string[] = []
+): boolean => {
+  if (item.hidden) return false;
+  if (!item.enableWhen && item.item) {
+    // This is a group. Only show it if some children are enabled.
+    return item.item.some((i) => isEnabled(i, disabledLinkIds));
+  }
+  return !disabledLinkIds.includes(item.linkId);
+};
+
+export const isShown = (item: FormItem, disabledLinkIds: string[] = []) => {
+  if (
+    !isEnabled(item, disabledLinkIds) &&
+    item.disabledDisplay !== DisabledDisplay.Protected
+  )
+    return false;
+
+  return true;
+};
+
+const DynamicViewFields: React.FC<Props> = ({
+  definition,
+  bulk,
+  itemMap,
+  // autofillDependencyMap, // { linkId => array of Link IDs that depend on it for autofill }
+  enabledDependencyMap, // { linkId => array of Link IDs that depend on it for enabled status }
+  horizontal = false,
+  pickListRelationId,
+  values,
+  setValues,
+  disabledLinkIds,
+  setDisabledLinkIds,
+  ...fieldsProps
+}) => {
+  const updateDisabledLinkIds = useCallback(
+    (changedLinkIds: string[], localValues: any) => {
+      setDisabledLinkIdsBase(changedLinkIds, localValues, setDisabledLinkIds, {
+        enabledDependencyMap,
+        itemMap,
+      });
+    },
+    [itemMap, enabledDependencyMap, setDisabledLinkIds]
+  );
+
+  const adjustValue: AdjustValueFn = useCallback(
+    (input) => {
+      const { linkId, value } = input;
+      setValues((currentValues) => {
+        const newValues = { ...currentValues };
+        newValues[linkId] = value;
+        updateDisabledLinkIds([linkId], newValues);
+
+        return newValues;
+      });
+    },
+    [updateDisabledLinkIds, setValues]
+  );
+
+  // Recursively render an item
+  const renderItem = (
+    item: FormItem,
+    nestingLevel: number,
+    props?: OverrideableDynamicFieldProps,
+    renderFn?: (children: ReactNode) => ReactNode
+  ) => {
+    const isDisabled = !isEnabled(item, disabledLinkIds);
+    if (isDisabled && item.disabledDisplay !== DisabledDisplay.Protected)
+      return null;
+    if (bulk && item.serviceDetailType === ServiceDetailType.Client)
+      return null;
+
+    if (item.type === ItemType.Group) {
+      return (
+        <DynamicViewGroup
+          item={item}
+          key={item.linkId}
+          nestingLevel={nestingLevel}
+          renderChildItem={(item, props, fn) =>
+            renderItem(item, nestingLevel + 1, props, fn)
+          }
+          values={values}
+          {...fieldsProps}
+        />
+      );
+    }
+
+    const itemComponent = (
+      <DynamicViewField
+        key={item.linkId}
+        item={item}
+        value={isDisabled ? undefined : values[item.linkId]}
+        nestingLevel={nestingLevel}
+        horizontal={horizontal}
+        pickListRelationId={pickListRelationId}
+        adjustValue={adjustValue}
```

Comment:
@gigxz So if I don't run `setDisabledLinkIds` with updated values when the pick list values are fetched, then fields will be visible that should not be. I'm all for storing it as state, but I also either need to run `setDisabledLinkIds` with the updated values, or we'll need to make it so that the disabled link IDs initial state can be derived correctly without needing pick list data loaded. I'd definitely prefer the latter if possible.

It looks like maybe the `enabledDependencyMap` part of the logic only gets fired when values are updated, and not actually during the initial state calculation. Though if I call it with old (pre-fetch) values then it doesn't hide the fields that should be hidden, so it looks like it also has to do with the initial values.

What do you think? You know the plumbing in there better than I do

Reply:
Oh I see, that's right. For Prior Living Situation on the Intake assessment, we have a conditional that depends on the "groupLabel" of the pick list option, which we only know once we've fetched it. I don't love that, because it presents as the dependent items appearing and then disappearing (both on the view and edit pages). But I don't think that's worth spending more time on. I'll merge this in as is. Thanks!